### PR TITLE
Stop building imports from a config-supplied model_type

### PR DESCRIPTION
**Pull request by danielhanchen**
`model_type` is read straight out of a downloaded `config.json` and then interpolated into code.

`unsloth_compile_transformers` builds `transformers.models.{model_type}.modeling_{model_type}` and `exec`'d an import of it, and `create_new_function` builds the compiled cache filename `unsloth_compiled_module_{model_type}.py` from the same string. A config carrying `"model_type": "llama'); import os; os.system(...)"` reached that `exec`.

### What this does

1. `hf_utils.py`: validate `model_type` against `[a-z0-9_]+` right after the existing `-`, `/`, `.` normalisation. This is the single choke point, since every caller in unsloth (`models/loader.py:766` and `models/loader.py:1560`) takes its `model_type` from `get_transformers_model_type`.
2. `compiler.py`: replace `exec(f"import {model_location}", globals())` plus `eval(model_location)` with `importlib.import_module`. The follow up `exec("import transformers", globals())` is required, not cosmetic: roughly 20 later call sites do `eval(f"{model_location}.{module}")`, and `create_standalone_class` does so from a different function, so `transformers` has to stay bound in module globals. `ModuleNotFoundError` behaviour is unchanged, including the case where the modeling file's own optional dependency is missing.
3. `empty_model.py`: replace `exec(f"{prefix}{key} = val")` in `set_additional_modules` with a dotted module path walk. The old prefixes `new_` and `new_model.` were a string trick, `"new_" + "model.visual.x"` spelling `new_model.visual.x`, so the two candidates meant "key with the leading `model.` consumed" and "key as is". Both candidates are preserved in the same order.

### Behaviour change worth flagging

Keys with a numeric segment such as `blocks.0.weight` are a syntax error for `exec`, so they were always silently skipped and that parameter was left unset. The walk handles them via `obj[int(part)]`, so those weights now get assigned. The old bare `except: continue` also hid genuine failures; failures now raise per candidate and a key that resolves under neither candidate produces a warning naming both errors.

### Testing

- All 5489 model types this transformers version ships (`CONFIG_MAPPING_NAMES` plus `dir(transformers.models)`), plus known remote code types such as `nemotron_h`, `chatglm`, `minicpmv`, `MiniCPM-V-2_6`, `gemma3_text`: none rejected by the new guard.
- Injection payloads (`llama'); import os; os.system(...)`, newline, `;`, `,`, NUL, `#`, ` as x; import os`, empty string) rejected. Path traversal such as `../../../etc/passwd` cannot survive, since `/` and `.` are already rewritten to `_` before the guard.
- `unsloth_compile_transformers("llama")` still patches the modeling file and reports `supports_sdpa=[True]`.
- 4-bit LoRA training on Llama-3.2-1B-Instruct, 5 steps plus eval plus generate: loss 1.5386 to 1.5061, eval loss 1.6985 to 1.6020.
- LoRA adapter save then reload through `FastLanguageModel.from_pretrained`, which exercises the `PeftConfig` branch of `get_transformers_model_type`.
- vLLM path (`fast_inference=True`), which is the only caller of `set_additional_modules`: Llama-3.2-1B-Instruct loads and generates, and Gemma-3-4B-it assigns all 8 of its `additional_keys` (vision tower embeddings, post layernorms, multi modal projector, language model norm) with no warnings and no parameters left on meta.
- New tests: `tests/security/test_model_type_import_guard.py` and `tests/test_set_additional_modules_paths.py`, 22 passed.

Verified on transformers 4.57.6, TRL 0.25.1, torch 2.9.1+cu128, vLLM 0.15.1. A transformers 5.x and TRL 1.x matrix run is in progress and I will follow up here with the results.

Unrelated issue found while testing, not caused by this change and reproduced identically on `main`: Qwen2.5-VL-3B bnb-4bit through vLLM 0.15.1 fails in vLLM's own `qwen2.py load_weights` on `assert param_data.shape == loaded_weight.shape`, before any unsloth code runs.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Following up on the version matrix promised above. Ran each cell in an isolated venv against this branch, torch 2.9.1+cu128 throughout, and each cell runs both the guard/compile test and a 5 step 4-bit LoRA train plus generate on Llama-3.2-1B-Instruct.

| transformers | trl | model types accepted | rejected | compile path | LoRA smoke |
|---|---|---|---|---|---|
| 5.15.1 | 1.10.0 | 5783 | 0 | ok, supports_sdpa=[True] | exit 0, train_loss 1.665 |
| 5.0.0 | 1.10.0 | 4601 | 0 | ok, supports_sdpa=[True] | exit 0, train_loss 1.665 |
| 4.57.6 | 0.22.2 | 5489 | 0 | ok, supports_sdpa=[True] | exit 0, train_loss 1.6654455 |
| 4.57.6 | 0.29.1 | 5489 | 0 | ok, supports_sdpa=[True] | exit 0, train_loss 1.6654455 |

No cell was skipped. Points worth noting:

- Zero legitimate `model_type` rejections on any transformers version. The accepted counts differ because each version ships a different set, and the regex covers all of `CONFIG_MAPPING_NAMES` plus `dir(transformers.models)` plus the community list after the `-`, `/`, `.` to `_` normalisation.
- The `importlib` change holds on 5.x. The test asserts both `modeling_llama.__UNSLOTH_PATCHED__` and that `transformers` is still bound in `compiler.__dict__`, so the added `exec("import transformers", globals())` does replace the side effect the old `exec(f"import {model_location}")` provided for the downstream `eval()` call sites. The `ModuleNotFoundError` early return path is intact too.
- Identical `train_loss` across cells, which is the point: these changes are behaviourally inert for training.

Two environment issues showed up during the matrix and neither is related to this branch: installing unsloth with `--no-deps` leaves `PIL` missing for `unsloth_zoo/vision_utils.py`, and `xformers==0.0.34` silently pulls torch up to 2.10.0, so the cells were run without xformers on the pinned torch.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/hf_utils.py [hunk 1]
```diff
@@ -215,6 +215,9 @@ def find(data, target_key):
         model_type = model_type.replace("-", "_")
         model_type = model_type.replace("/", "_")
         model_type = model_type.replace(".", "_")
+        # model_type is interpolated into an import path, so it must be a plain module name
+        if not re.fullmatch(r"[a-z0-9_]+", model_type):
+            raise ValueError(f"Unsloth: Invalid model_type {model_type!r} in config.")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/hf_utils.py:225]: Correct, and reproduced before fixing: AutoConfig.for_model("dbrx") walks to ['dbrx', '', ''], got_ocr2 to ['got_ocr2', 'qwen2', ''] and qwen3_omni_moe to a list with two empty entries, so the guard raised on all three. The root cause is general rather than per family: PretrainedConfig.model_type defaults to "" (transformers/configuration_utils.py), so any nested sub-config that does not override it serialises as an empty sentinel. Fixed in 764d37c9 by skipping empty values before the guard, and raising the existing cannot determine error if nothing but sentinels is left. Regression tests cover all three families plus the sentinels-only case.

**Comment by danielhanchen** [comment]
@codex review
